### PR TITLE
Propagate fake RuntimeErrors to Dynamo handlers

### LINKED ISSUE #148475: Dynamo replaces exception by hard error in `run_node`
### 🐛 Describe the bug

While working on PR [#146500](https://github.com/pytorch/pytorch/pull/146500), I noticed that some tests assert that a PyTorch function raises an exception for certain inputs. One example is `TestTorchDeviceTypeCPU.test_broadcast_fn_ge_cpu`.

Below is a minimal reproducer of what most of the failing tests attempt to do:

```python
import torch

@torch.compile(backend='eager')
def fn(t):
    t0 = torch.randn(2)
    try:
        t.expand_as(t0)
    except RuntimeError:
        return t.sin()
    return t.cos()

t = torch.randn(2, 3)
y = fn(t)
print(y)
```

The call `t.expand_as(t0)` raises `RuntimeError('expand: the requested shape has too few dimensions!')`. However, the user code never gets a chance to handle this exception because it is replaced by a hard error ([`TorchRuntimeError`](https://github.com/pytorch/pytorch/blob/7fcbaff206d1626353b414f433110de2dc9d3f48/torch/_dynamo/utils.py#L3217)).

In theory, we could replace this hard error with something Dynamo can process, but doing so would incorporate the expand_as call into the computation graph:

```diff
diff --git a/torch/_dynamo/utils.py b/torch/_dynamo/utils.py
index 2f7f5b14d0b..865539fdcda 100644
--- a/torch/_dynamo/utils.py
+++ b/torch/_dynamo/utils.py
@@ -3214,7 +3214,8 @@ def get_fake_value(node, tx, allow_non_graph_fake=False):
                 hints=[],
             )

-        raise TorchRuntimeError(str(e)).with_traceback(e.__traceback__) from None
+        from .exc import raise_observed_exception
+        raise_observed_exception(type(e), tx)

     if not allow_non_graph_fake:
         _ = pytree.tree_map_only(
```

```python
class GraphModule(torch.nn.Module):
   def forward(self, L_t_: "f32[2, 3][3, 1]cpu"):
       l_t_ = L_t_

        # File: /home/guilhermeleobas/git/pytorch/test.py:6 in fn, code: t0 = torch.randn(2)
       t0: "f32[2][1]cpu" = torch.randn(2)

        # File: /home/guilhermeleobas/git/pytorch/test.py:8 in fn, code: t.expand_as(t0)
       expand_as = l_t_.expand_as(t0);  t0 = expand_as = None

        # File: /home/guilhermeleobas/git/pytorch/test.py:10 in fn, code: return t.sin()
       sin: "f32[2, 3][3, 1]cpu" = l_t_.sin();  l_t_ = None
       return (sin,)
```

If we include this in the graph, the same error would occur at runtime, which is not ideal. 

Could Dynamo handle this kind of pattern in some way? Or this is probably something that can not be supported.

cc @zou3519 

### Versions

main branch

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185776

Dynamo computes example values for newly traced FX nodes by running them
under fake tensor mode. If that fake execution raised a RuntimeError,
get_fake_value always wrapped it as a hard TorchRuntimeError. That meant
user code inside `try/except RuntimeError` never got a chance to handle
errors such as invalid `expand_as` shapes.

A direct conversion to an observed exception is not sufficient because the
failed FX node has already been inserted into the graph. If left in place,
the same invalid op can survive into the compiled graph and fail at runtime.

Fix this by detecting when fake execution fails while the current translator
or an inlined parent translator is in an active exception region. In that
case, remove the failed FX node and raise Dynamo's observed RuntimeError so
existing exception-table handling can route control to the user's handler.
When there is no active handler, keep the existing TorchRuntimeError path.

I considered always converting fake RuntimeErrors to observed exceptions,
but that would change no-handler behavior and could hide ordinary fake tensor
failures. Limiting the conversion to active exception regions keeps the
existing diagnostics outside user exception handling.

Fixes #148475
Generated by my agent

Benchmark Results:
- Baseline valid `expand_as` compile+first-run: mean 18.723 ms, median 18.645 ms over 30 iterations.
- Patched valid `expand_as` compile+first-run: mean 20.024 ms, median 19.767 ms over 30 iterations.
- Baseline handled-error reproducer could not complete; it raised `torch._dynamo.exc.TorchRuntimeError`.
- Patched handled-error reproducer compile+first-run: mean 22.335 ms, median 21.711 ms over 30 iterations.

Test Plan:
- python test/dynamo/test_exceptions.py -k fake_tensor_runtime_error
- python test/dynamo/test_logging.py LoggingTests.test_dynamo_error
- lintrunner -a
- git diff --cached --check

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @jataylo @azahed98

Accuracy CSV note:
- The expected CSV changes are for CUDA A10G CI training accuracy rows where the benchmark reported `eager_two_runs_differ` before producing any Dynamo graph, then separately reported the graph-break count improving to zero. In the pre-CSV-update PR CI run 27695826533, `tf_efficientnet_b0` hit `eager_two_runs_differ` with `Dynamo produced 0 graphs covering 0 ops with 0 graph breaks` in job 81927321415, and `mnasnet1_0` / `mobilenet_v2` did the same in job 81927321301. The same logs then reported graph-break improvements from 6-7 down to 0 for those rows, which is why the expected graph-break column also changed to 0.
- I also checked local `main` on this B200 host with CI-style `--training --amp --inductor --accuracy --no-translation-validation` runs: `mnasnet1_0` and `mobilenet_v2` passed locally with the old graph-break counts, so the eager nondeterminism appears A10G-CI-specific rather than reproducible on this B200 environment.

Additional local checks from the latest update:
- PYTHONPATH=/data/users/jansel/pytorch-issue-fixer-2/pytorch:/data/users/jansel/torchvision LD_LIBRARY_PATH=/data/users/jansel/env/lib /data/users/jansel/env/bin/python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --print-compilation-time --device cuda --training --amp --inductor --no-translation-validation --only mnasnet1_0 --output /tmp/pytorch-148475-main-mnasnet1_0.csv
- PYTHONPATH=/data/users/jansel/pytorch-issue-fixer-2/pytorch:/data/users/jansel/torchvision LD_LIBRARY_PATH=/data/users/jansel/env/lib /data/users/jansel/env/bin/python benchmarks/dynamo/torchbench.py --ci --accuracy --timing --explain --print-compilation-time --device cuda --training --amp --inductor --no-translation-validation --only mobilenet_v2 --output /tmp/pytorch-148475-main-mobilenet_v2.csv

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the new dynamo_wrapped CI failures by narrowing fake RuntimeError propagation to real Python exception handlers and removing two now-obsolete Dynamo expected-failure markers. Targeted local checks pass for the fake tensor exception tests and test_broadcast_fn_add_cpu under PYTORCH_TEST_WITH_DYNAMO=1.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the new dynamo_wrapped addcdiv CI failure by removing the now-stale expected-failure marker for test_broadcast_fn_addcdiv_cpu. I also handled the review robustness feedback by initializing the handler scan index and preserving RuntimeError subclasses when fake execution routes an observed exception to user handlers. Targeted local checks pass for the fake tensor exception tests, add/addcdiv under PYTORCH_TEST_WITH_DYNAMO=1, py_compile, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I addressed the new dynamo_wrapped addcmul CI failure by removing the now-stale expected-failure marker for test_broadcast_fn_addcmul_cpu. While doing that, review found fake RuntimeError exception contents could be unsafely observed from user handlers, so I tightened the fix: type matching still works, but inspecting fake-derived exception messages/args now graph-breaks instead of compiling against fake metadata. Targeted checks pass for the fake tensor exception tests, the addcmul dynamo_wrapped test, py_compile, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the latest Dr. CI failures by skipping the sys.exception inspection test on Python < 3.11, where that API does not exist, and by removing the now-stale Dynamo expected-failure marker for test_broadcast_fn_copy_cpu. Targeted local checks pass for the fake tensor exception tests, the copy broadcast test under PYTORCH_TEST_WITH_DYNAMO=1, py_compile, and lintrunner -a.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR for the new dynamo_wrapped div failure by removing the stale TestTorchDeviceTypeCPU.test_broadcast_fn_div_cpu expected-failure marker after the targeted test passed. I also tightened the fake RuntimeError path so only user-visible fake RuntimeErrors reach real user handlers, while fake tensor internal RuntimeErrors stay on the TorchRuntimeError path and unsafe fake exceptions cannot be inspected or reconstructed.

**Comment by jansel** [comment]
My agent says I addressed the latest Dr. CI dynamo_wrapped failures by removing the now-stale expected-failure marker for TestTorchDeviceTypeCPU.test_broadcast_fn_lerp_cpu after the targeted PYTORCH_TEST_WITH_DYNAMO=1 run passed locally. I also reran the fake tensor RuntimeError Dynamo tests, py_compile, and lintrunner -a.

**Comment by jansel** [comment]
My agent says I addressed the latest Dr. CI dynamo_wrapped failures by removing the now-stale expected-failure marker for TestTorchDeviceTypeCPU.test_broadcast_fn_map2_cpu after the targeted PYTORCH_TEST_WITH_DYNAMO=1 run passed locally. I also reran the fake tensor RuntimeError Dynamo tests and lintrunner -a.

**Comment by jansel** [comment]
My agent says I addressed the latest Dr. CI dynamo_wrapped failures by removing stale expected-failure markers for test_broadcast_fn_map_cpu, masked_fill, masked_scatter, mul, and sub after targeted runs and the broader TestTorchDeviceTypeCPU -k test_broadcast suite passed under PYTORCH_TEST_WITH_DYNAMO=1. I also triaged the shufflenet inductor_torchbench failure as eager_two_runs_differ on the merge base, so it does not look caused by this PR. Local checks passed for the fake tensor RuntimeError tests, lintrunner -a, py_compile, and git diff --cached --check.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR for the latest Dr. CI failures and review finding. The inductor benchmark jobs were failing because this PR eliminates graph breaks for tf_efficientnet_b0, mnasnet1_0, and mobilenet_v2, leaving those models at eager_two_runs_differ with zero graph breaks, so I updated the expected CSV baselines accordingly. I also fixed the with-inside-outer-try/except gap by following WITH_EXCEPT_START re-raise chains and using internal traceback access during Dynamo's context-manager exception bookkeeping. Targeted fake RuntimeError tests, synthetic benchmark checks, py_compile, lintrunner -a, and review pass.

**Comment by jansel** [comment]
My agent says I updated the PR description with the A10G CI evidence for the three accuracy CSV rows. The logs show the eager_two_runs_differ status was emitted before any Dynamo graph was produced, followed separately by graph-break improvements to zero; local B200 main checks for mnasnet1_0 and mobilenet_v2 did not reproduce the A10G eager nondeterminism.

**Comment by guilhermeleobas** [comment]
There's already one PR opened for this: https://github.com/pytorch/pytorch/pull/182972

**Comment by jansel** [comment]
My agent says I checked #182972 before deciding what to do here. It is still open with requested changes and merge conflicts, and #185776 is not a pure duplicate: it adds fake-internal error filtering, only routes fake RuntimeErrors when a real user handler is active, and prevents compiled code from inspecting fake-derived exception contents. I am keeping this open for review unless maintainers decide those pieces should instead be ported into #182972.

**Comment by morrison-turnansky** [comment]
_has_active_exception_handler is very similar to my original approach of using _has_exception_handler in https://github.com/pytorch/pytorch/pull/182972.  I was told to abandon this approach and instead use the logic of bubbling exceptions already in dynamo.  My PR currently follows this second approach.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I addressed the feedback about avoiding the abandoned active-handler scan approach. The PR now routes eligible fake RuntimeErrors through Dynamo's existing observed-exception bubbling path, lets real user handlers catch them, and converts uncaught fake RuntimeErrors back to TorchRuntimeError. I kept the PR's fake-internal RuntimeError filtering, failed-node removal, and unsafe-inspection guard, and added coverage for preserving custom __str__ on safe user exceptions. Local checks passed: py_compile, fake_tensor_runtime_error tests, exception str/repr tests, PYTORCH_TEST_WITH_DYNAMO=1 TestTorchDeviceTypeCPU -k test_broadcast, lintrunner -a, diff checks, and a fresh review subagent pass.

**Comment by jansel** [comment]
My agent says I updated the PR after the latest Claude review and Dr. CI failures. This push removes the dead var_getattr fallbacks, documents the fake/internal RuntimeError contract, preserves uncaught fake FX-node diagnostics, handles UDOV-wrapped fake errors, guards __exit__ traceback inspection, avoids repr/str on fake non-string RuntimeError args, restores the three A10G expected CSV rows to pass with their observed graph-break counts, and removes stale tensorinv expected-failure markers. Local targeted tests and lintrunner passed; the tensorinv linalg target was blocked locally by a torchvision::nms import mismatch.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I updated the PR for the latest Dr. CI and review findings. The tensorinv dynamo_expected_failure markers are restored because CI showed those tests still fail under dynamo_wrapped; fake-internal MissingOpProfile is no longer catchable by user RuntimeError handlers; unsafe fake exception diagnostics now avoid user args/repr/str and metaclass __name__ hooks; recursive container formatting is cycle-safe; and yield-from throw/close paths preserve generator exhaustion after fake RuntimeErrors. Local focused fake RuntimeError tests, py_compile, lintrunner -a, and review subagents passed. The local tensorinv target is still blocked before test execution by the torchvision::nms import mismatch.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased this atop the competing fix and refocused it on safe exception routing. The update removes failed FX nodes, routes only audited or explicit RuntimeErrors, keeps custom fake, mutable Meta, and real-propagation failures on the hard-error path, prevents fake exception inspection, narrows nonstrict output handling, fixes uint8 mask metadata, and restores 15 expected failures removed by the broad-routing change. Focused tests, adversarial reproducers, lintrunner -a, and an independent maximum-depth review all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and addressed the 24 new Dr. CI failures. RuntimeError routing now uses scoped private-type provenance only for audited core validation, while custom fake, propagated-real, and arbitrary subclass errors remain hard without mutating user exceptions; auxiliary real errors retain their original messages. I also made custom-op diagnostics hook-safe, corrected the uint8 indexing test, removed the now-passing NumPy expected-failure marker, and restored the Python 3.14 PackedSequence/Hessian markers that the broad-routing change had removed. The 45 focused exception tests, exact CI reproducers, nonstrict-output tests, logging test, lintrunner, and a fresh review all pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto current main and fixed all six related Dr. CI failures: preserved CPython traceback wording and restored expected failures for the unaudited FX and PackedSequence fake errors. Fresh maximum-rigor review also found and I fixed three provenance leaks in mutable prim meta implementations, propagate_real_tensors auxiliary kernels, and unsafe fallback kernels; these synthetic or input-dependent executions now remain on the hard-error path, with regressions for each prior silent wrong-code case. The full 143-test exception suite, targeted expected-failure checks, lintrunner, py_compile, and diff checks pass.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
